toolchain/base/mem_usage.h

```diff
@@ -0,0 +1,113 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_BASE_MEM_USAGE_H_
+#define CARBON_TOOLCHAIN_BASE_MEM_USAGE_H_
+
+#include <cstdint>
+
+#include "common/map.h"
+#include "common/set.h"
+#include "llvm/ADT/STLFunctionalExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "toolchain/base/yaml.h"
+
+namespace Carbon {
+
+// Types supporting memory usage tracking should implement:
+//
+//   // Collects memory usage of members.
+//   auto CollectMemUsage(llvm::StringRef label, AddMemUsageFn add_mem_usage)
+//       const -> void;
+//
+// The label should be concatenated with any child labels using MemUsageLabel in
+// order to reflect allocation structure.
+//
+// The arguments for AddMemUsageFn are the label and byte size. It should be
+// called once per tracked size.
+class MemUsage {
+ public:
+  auto Add(std::string label, int64_t used, int64_t reserved) -> void {
+    mem_usage_.push_back({std::move(label), used, reserved});
+  }
+
+  auto Add(std::string label, const llvm::BumpPtrAllocator& allocator) -> void {
+    mem_usage_.push_back({std::move(label), allocator.getBytesAllocated(),
+                          allocator.getTotalMemory()});
+  }
+
+  template <typename KeyT, typename ValueT, ssize_t SmallSize,
+            typename KeyContextT>
+  auto Add(std::string label, Map<KeyT, ValueT, SmallSize, KeyContextT> map,
+           KeyContextT key_context = KeyContextT()) -> void {
+    auto bytes = map.ComputeMetrics(key_context).storage_bytes;
+    mem_usage_.push_back({std::move(label), bytes, bytes});
+  }
+
+  template <typename KeyT, ssize_t SmallSize, typename KeyContextT>
+  auto Add(std::string label, Set<KeyT, SmallSize, KeyContextT> map,
+           KeyContextT key_context = KeyContextT()) -> void {
+    auto bytes = map.ComputeMetrics(key_context).storage_bytes;
+    mem_usage_.push_back({std::move(label), bytes, bytes});
+  }
+
+  // Adds memory usage of an array's data. This ignores the possible overhead of
+  // a SmallVector's in-place storage; if it's used, it's going to be tiny
+  // relative to scaling memory costs.
+  //
+  // This uses SmallVector in order to get proper inference for T, which
+  // ArrayRef misses.
+  template <typename T, unsigned N>
+  auto Add(std::string label, const llvm::SmallVector<T, N>& array) -> void {
+    Add(std::move(label), array.size_in_bytes(), array.capacity_in_bytes());
+  }
+
+  template <typename T>
+  auto Collect(llvm::StringRef label, const T& arg) -> void {
+    arg.CollectMemUsage(*this, label);
+  }
```

Comment:
What is the advantage of using this function instead of calling `CollectMemUsage` directly?